format/ComputeIR.fbs

```diff
@@ -0,0 +1,241 @@
+  /// "filter" relation's output.
+  filter_expression: Expression (required);
+}
+
+/// The contents of Relation.options will be ProjectOptions
+/// if Relation.relation_name = "project"
+table ProjectOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "project" relation's output.
+  expressions: [Expression] (required);
+}
+
+/// The contents of Relation.options will be AggregateOptions
+/// if Relation.relation_name = "aggregate"
+table AggregateOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "aggregate" relation's output.
+  aggregations: [Expression] (required);
+  /// Keys by which `aggregations` will be grouped.
+  keys: [Expression];
+}
+
+/// The contents of Relation.options will be JoinOptions
+/// if Relation.relation_name = "join"
+table JoinOptions {
+  /// The expression which will be evaluated against rows from each
+  /// input to determine whether they should be included in the
+  /// "join" relation's output.
+  on_expression: Expression (required);
+  join_kind: string;
+}
+
+/// Whether lesser values should precede greater or vice versa,
+/// also whether nulls should preced or follow values.
+enum Ordering : uint8 {
+  ASCENDING_THEN_NULLS,
+  DESCENDING_THEN_NULLS,
+  NULLS_THEN_ASCENDING,
+  NULLS_THEN_DESCENDING
+}
+
+table SortKey {
+  value: Expression (required);
+  ordering: Ordering = ASCENDING_THEN_NULLS;
+}
+
+/// The contents of Relation.options will be OrderByOptions
+/// if Relation.relation_name = "order_by"
+table OrderByOptions {
+  /// Define sort order for rows of output.
+  /// Keys with higher precedence are ordered ahead of other keys.
+  keys: [SortKey] (required);
+}
+
+/// The contents of Relation.options will be LimitOptions
+/// if Relation.relation_name = "limit"
+table LimitOptions {
+  /// Set the maximum number of rows of output.
+  count: long;
+}
+
+/// The contents of Relation.options will be CommonOptions
+/// if Relation.relation_name = "common"
+table CommonOptions {
+  /// Commons (CTEs in SQL) allow assigning a name to a stream
+  /// of data and reusing it, potentially multiple times and
+  /// potentially recursively.
+  name: string;
+}
+
+/// The contents of Relation.options will be UnionOptions
+/// if Relation.relation_name = "union"
+table UnionOptions {
+  /// For simplicity, all rows from any input to a "union" relation
+  /// will always be concatenated into a single output- establishing
+  /// uniqueness of output rows is deferred to other relations.
+}
+
+/// The contents of Relation.options will be LiteralOptions
+/// if Relation.relation_name = "literal"
+table LiteralOptions {
+  /// Batches of rows in this literal.
+  batches: [RecordBatch] (required);
+}
+
+/// A specification of a query.
+table Plan {
+  /// One or more output relations.
+  sinks: [Relation] (required);
+
+  /// Dictionary batches which may be referenced by Literals.
+  dictionary_batch: [DictionaryBatch];
```

Comment:
Can you elaborate a tiny bit on why dictionary batches are in `Plan` and not in `LiteralOptions`?

Reply:
Some `Literal`s (expressions here, not literal relations) may be dictionary encoded, in which case they need to refer into a dictionary batch (this organization is inherited from the arrow IPC format). I could instead move this back to a member of LiteralOptions, but I'd need to add it as a member of Literal (expression) as well